AWX: Add `termination_grace_period_seconds`

stanislav-zaprudskiy opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

During deployment restarts or new rollouts, when old ReplicaSet Pods are being terminated, the corresponding jobs which are managed (executed or controlled) by old AWX Pods may end up in `Error` state as there is no mechanism to transfer them to the newly spawned AWX Pods. To work around the problem one could set `termination_grace_period_seconds` in AWX spec proposed by this PR, which does the following:

* It sets the corresponding [`terminationGracePeriodSeconds`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) Pod spec of the AWX Deployment to the value provided

  > The grace period is the duration in seconds after the processes running in the pod are sent a termination signal and the time when the processes are forcibly halted with a kill signal

* It adds a [`PreStop`](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution) hook script, which will keep AWX Pods in terminating state until it finished, up to `terminationGracePeriodSeconds`.

  > This grace period applies to the total time it takes for both the PreStop hook to execute and for the Container to stop normally

  While the hook script just waits until the corresponding AWX Pod (instance) no longer has any managed jobs, in which case it finishes with success and hands over the overall Pod termination process to normal AWX processes.

<!---
If you are fixing an existing issue, please include "fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
related [#11526](https://github.com/ansible/awx/issues/11526)
related [#12326](https://github.com/ansible/awx/issues/12326)

requires https://github.com/ansible/awx/pull/13445

rebased onto https://github.com/ansible/awx-operator/pull/1222

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### ADDITIONAL INFORMATION

There are two `PreStop` hook scripts utilized by AWX Pod containers, all of which communicate with each other through files in a shared volume:
* `termination-master` - runs a so called "handler" command (`awx-manage disable_instance --wait --retry=inf`) until it finishes, or until it receives a "signal" (through files) from other instances of hook scripts. In case of successful handler termination it creates a "termination marker", signalling other instances of hook scripts that they could finish and thus proceed to the remaining container processes termination.
* `termination-waiter` - continously runs in a loop and checks for presence of "signals" from other hook scripts. Apart from checking for "termination marker", it also checks if the corresponding "heartbeat" file was recently updated (by `master`) and fails otherwise.

Out of 4 current containers only `awx-task` has `termination-master` as its hook, while others are `waiter`s (including `redis`).

Some script logic is added to account for [Hook delivery guarantees](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-delivery-guarantees).

The reason to have all scripts outside of `awx` image is that the scripts are also executed by `redis`.

To play with scripts one could use the following:

<details>
  <summary>add `roles/installer/files/pre-stop/termination-handler`</summary>

  ```bash
#!/usr/bin/env bash

source $(dirname "$0")/termination-env > /dev/null

thrsh=${PRE_STOP_HANDLER_RND_THRESHOLD-98}

while true; do
  rnd=$(( $RANDOM % 100 ))
  sleep="${PRE_STOP_HANDLER_SLEEP:-$(( $heartbeat_threshold + ($RANDOM % 2) * ($RANDOM % 2) * ($RANDOM % 2) ))}"
  printf "%s: will sleep %ss if %s -ge %s\n" \
    "$FOO" \
    "$sleep" \
    "$thrsh" \
    "$rnd"
  if [[ $rnd -lt $thrsh ]]; then
    sleep $sleep
  else
    # exit 116
    break
  fi
done
  ```
</details>


<details>
  <summary>run test commands, manipulating environment as needed</summary>

  ```console
temp_dir=$(mktemp -d) \
&& export \
  PRE_STOP_HEARTBEAT_THRESHOLD=2 \
  PRE_STOP_HANDLER_SLEEP= \
  PRE_STOP_HANDLER="bash -c \"FOO=bar $PWD/roles/installer/files/pre-stop/termination-handler\"" \
  PRE_STOP_MARKER_FILE=$temp_dir/.termination_marker \
  PRE_STOP_HEARTBEAT_FILE=$temp_dir/.heartbeat \
  PRE_STOP_BAILOUT_FILE=$temp_dir/.bailout \
  PRE_STOP_STDOUT=/dev/stdout \
  PRE_STOP_STDERR=/dev/stderr; \
{ ./roles/installer/files/pre-stop/termination-master & }; \
./roles/installer/files/pre-stop/termination-waiter
  ```
</details>

<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
-->

<!--- Paste verbatim command output below, e.g. before and after your change -->

Container logs output (note that redis logs aren't available as it does not allow to stream to PID1's stdout and stderr):

<details>
  <summary>formatted for brevity</summary>

  ```
[pod/awx-5558dbb966-vndjw/awx-web] 2023-01-18T16:35:25.162581251Z [preStop.exec] [waiter] The hook has started: marker_file="/var/lib/pre-stop/.termination_marker" heartbeat_file="/var/lib/pre-stop/.heartbeat" bailout_file="/var/lib/pre-stop/.bailout" recheck_sleep="1" report_every="30"
[pod/awx-5558dbb966-vndjw/awx-ee] 2023-01-18T16:35:25.263185671Z [preStop.exec] [waiter] The hook has started: marker_file="/var/lib/pre-stop/.termination_marker" heartbeat_file="/var/lib/pre-stop/.heartbeat" bailout_file="/var/lib/pre-stop/.bailout" recheck_sleep="1" report_every="30"
[pod/awx-5558dbb966-vndjw/awx-ee] 2023-01-18T16:35:25.284689290Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-web] 2023-01-18T16:35:25.284809768Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T16:35:25.310203210Z [preStop.exec] [master] The hook has started: marker_file="/var/lib/pre-stop/.termination_marker" heartbeat_file="/var/lib/pre-stop/.heartbeat" bailout_file="/var/lib/pre-stop/.bailout" handler="bash -c "PYTHONUNBUFFERED=x awx-manage disable_instance --wait --retry=inf""
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T16:35:27.766206323Z [preStop.exec] [master] [handler] Instance ["awx-5558dbb966-vndjw"](http://localhost:30080/#/instances/79/details) has been disabled
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T16:35:27.780099825Z [preStop.exec] [master] [handler] 1/inf: Waiting 30s before the next attempt to see if the following instance' managed jobs have finished: ["Inventory Sync Baremetal"](http://localhost:30080/#/jobs/playbook/357)
[pod/awx-5558dbb966-vndjw/awx-ee] 2023-01-18T16:35:55.706871333Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-web] 2023-01-18T16:35:55.770965479Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T16:35:57.798660631Z [preStop.exec] [master] [handler] 2/inf: Waiting 30s before the next attempt to see if the following instance' managed jobs have finished: ["Inventory Sync Baremetal"](http://localhost:30080/#/jobs/playbook/357)
...
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T17:15:31.023173580Z [preStop.exec] [master] [handler] 81/inf: Waiting 30s before the next attempt to see if the following instance' managed jobs have finished: ["Inventory Sync Baremetal"](http://localhost:30080/#/jobs/playbook/357)
[pod/awx-5558dbb966-vndjw/awx-ee] 2023-01-18T17:15:43.643723833Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-web] 2023-01-18T17:15:46.230869073Z [preStop.exec] [waiter] Waiting for the marker file to be accessible...
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T17:16:01.033448017Z [preStop.exec] [master] [handler] Done waiting for instance' managed jobs to finish!
[pod/awx-5558dbb966-vndjw/awx-task] 2023-01-18T17:16:01.381330223Z [preStop.exec] [master] Done! Generating the marker file allowing to proceed to termination
[pod/awx-5558dbb966-vndjw/awx-ee] 2023-01-18T17:16:01.732001541Z [preStop.exec] [waiter] The marker file found, exiting to proceed to termination
[pod/awx-5558dbb966-vndjw/awx-web] 2023-01-18T17:16:02.323429998Z [preStop.exec] [waiter] The marker file found, exiting to proceed to termination
  ```
</details>

stanislav-zaprudskiy committed:
AWX: Add `termination_grace_period_seconds`

stanislav-zaprudskiy committed:
Make AWX Pod variable to be calculated respecting `creationTimestamp` and `deletionTimestamp`

Do not consider Pods marked for deletion when calculating tower_pod to
address replicas scale down case - where normally Pods spawned recently
are being taken for removal. As well as the case when operator kicked
off but some old replicas are still terminating.

Respect `creationTimestamp` so to make sure that the newest Pod is taken
after Deployment application, in which case multiple RS Pods (from old
RS and new RS) could be running simultaneously while the rollout is
happening.

TheRealHaoLiu commented:
Hi @stanislav-zaprudskiy thank you for this really thorough PR, would u mind joining us and discuss (and maybe demo the PR) this topic on our monthly community meeting?

Agenda issue for the upcoming community meeting
https://github.com/ansible/awx/issues/13431

stanislav-zaprudskiy committed:
When applying Deployment wait up to (timeout * replicas)

There are cases when having a new Deployment may be taking above the
default timeout of 120s.
For instance, when a Deployment has multiple replicas, and each replica
starts on a separate node, and the Deployment specifies new images, then
just pulling these new images for each replica may be taking above the
default timeout of 120s.

Having the default time multiplied by the number of replicas should
provide generally enough time for all replicas to start

stanislav-zaprudskiy committed:
Make Deployment to be rolled out on CM and Secrets changes

With the previous approach, not all associated (mounted) CM/Secrets
changes caused the Deployment to be rolled out, but also the Deployment
could have been rolled out unnecessary during e.g. Ingress or Service
changes (which do not require Pod restarts).

Previously existing Pod removal (state: absent) was not complete as
other pods continued to exist, but also is not needed with this commit
change due to added Pods annotations.

The added Deployment Pod annotations now cause the new ReplicaSet
version to be rolled out, effectively causing replacement of the
previously existing Pods in accordance with the deployment `strategy`
(https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#deploymentstrategy-v1-apps,
`RollingUpdate`) whenever there is a change in the associated CMs or
Secrets referenced in annotations. This implementation is quite standard
and widely used for Helm workflows -
https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

stanislav-zaprudskiy committed:
Do not wait for a new Pod name after Deployment change

Proper waiting is already performed earlier during Deplyment{apply: yes, wait: yes} -
https://github.com/ansible-collections/kubernetes.core/blob/e6ac87409830b6c698b05dba875cca87d45ea761/plugins/module_utils/k8s/waiter.py#L27.

And also not every Deployment change produces new RS/Pods. For example,
changing Deployment labels won't cause new rollout, but will cause
`until` loop to be invoked unnecessarily (when replicas=1).

stanislav-zaprudskiy committed:
Fix lint warnings

Cl0udius commented:
Hi. 

Thanks for working on this topic!
As far as i can see, this would prevent jobs from failing in the case we shutdown control pods in a controlled way.
Is there also an idea how we can prevent jobs from failing when the control pod dies from a K8S host outage?
What i mean is: could there be also a way for the remaining control pods to takeover the jobs from died control pod?

TheRealHaoLiu commented:
https://github.com/ansible/awx-operator/pull/1218

here's our PR to split up web/task into separate deployments can u take a look and see how we need to react to your changes and if this impacts your implementation

stanislav-zaprudskiy committed:
Do not keep web container running during termination

We could let the web container terminate as usual, as there are no
reasons to keep it running as it doesn't participate in
job control. Additionally, it stops receiving traffic with the beginning
of termination

> At the same time as the kubelet is starting graceful shutdown, the
> control plane removes that shutting-down Pod from EndpointSlice (and
> Endpoints) objects where these represent a Service with a configured
> selector

@ https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination

stanislav-zaprudskiy commented:
> #1218
> 
> here's our PR to split up web/task into separate deployments can u take a look and see how we need to react to your changes and if this impacts your implementation

@TheRealHaoLiu,

AFAICS, the web container does not need to continue to be running during Pod termination, see https://github.com/ansible/awx-operator/pull/1193/commits/49d1f00dbd5c2cdd840333e9de03b3789f45de0e. Thus I would say in #1218, where the web part is split, the changes from this PR would only be relevant for the corresponding new `task` Deployment. I do not completely follow on the new `rsyslog` container though - but consider adding [the corresponding `termination-waiter` hook configuration](https://github.com/ansible/awx-operator/pull/1193/files#diff-29de64e6284e07a8a16b351b132d812618c7f4e3485cec09987c2374257b94fcR411-R422) for it as well, in case it needs to be running for the `task` container to save the data gracefully during shutdown.

TheRealHaoLiu commented:
Testing process for the PR 

Setup:
- OpenShift cluster 4.11.20
- AWX built from https://github.com/ansible/awx/pull/13445
- AWX operator built from this PR 

🟢 scenario 0:
- validate that with no job running and termination grace period set it does not increase the termination time of pod

scenario 1:
- `termination_grace_period_seconds` set to 120
- run 60 second job 
- 1 replicas
- kill awx
- validate that awx pod stay in termintation state 
- validate that awx UI is still accessible and the instance that we delete 

scenario 2:
- termination_grace_period_seconds` set to 60
- run 120 second jobs 
- 1 replicas
- validate the awx pod stay in termination state for only 60 second

scenario 3:
- termination_grace_period_seconds` set to 60
- run 120 second job
- 2 replicas
- kill pod that's NOT running a job 
- validate that awx pod is terminated without delay 

scenario 4 
- termination_grace_period_seconds` set to 60
- run 120 second job
- 2 replicas
- kill pod that's running the job 
- validate that awx pod is stay in terminated state for 60 second
